[`pyflakes`] Fix false positives for `__annotate__` (Py3.14+) and `__warningregistry__` (`F821`)

danparizher opened the pull request:
## Summary

Fixes #19970

ntBre commented on crates/ty_python_semantic/src/place.rs:
This change looks reasonable to me, but we'll want someone from the ty side to have a look too, maybe @AlexWaygood? I'm guessing they'd like you to have an mdtest for this as well. Possibly in [`moduletype_attrs.md`](https://github.com/astral-sh/ruff/blob/c3f1f0b9f1ed8e522a6d22c7d1670997c19183de/crates/ty_python_semantic/resources/mdtest/scopes/moduletype_attrs.md?plain=1#L1), but I'm not really sure.

ntBre commented on crates/ruff_python_stdlib/src/builtins.rs:
```diff
@@ -18,8 +18,10 @@ const IPYTHON_BUILTINS: &[&str] = &["__IPYTHON__", "display", "get_ipython"];
+    "__annotate__",
     "__builtins__",
     "__cached__",
+    "__warningregistry__",
```
I think it's correct to put `__warningregistry__` here because I don't think it's available in `builtins`, but I think we should create a separate `PY314_PLUS_BUILTINS` down below like the other version-specific groups for `__annotate__` since it was introduced in 3.14.

It appears not to _really_ matter since both `MAGIC_GLOBALS` and `python_builtins` are only called once each in the same place. But we may also need to update `is_python_builtin` since its comment says it's supposed to be kept in sync with `python_builtins`.

ntBre commented on crates/ruff_linter/src/rules/pyflakes/mod.rs:
```diff
@@ -920,6 +920,19 @@ mod tests {
+    }
+
+    #[test]
+    fn module_annotate_py314_plus() {
+        // __annotate__ is available starting in Python 3.14; ensure no F821 when enabled.
+        // Note: The helper uses default target version from settings; ensure CI targets latest.
+        flakes("__annotate__", &[]);
+    }
```
This doesn't seem to be testing what the test name suggests. As the comment says, the `flakes` test runner uses the `latest` version, which is still 3.13, so we'll need to write some custom test code here. We've been talking about switching this whole module over to the more common `test_path` and `assert_diagnostics` tests like most other rules, so I'd probably suggest just doing something like that here. One example of that:

https://github.com/astral-sh/ruff/blob/c3f1f0b9f1ed8e522a6d22c7d1670997c19183de/crates/ruff_linter/src/rules/ruff/mod.rs#L116-L125

We should also test that `__annotate__` is not available before 3.14. I think this test should actually be failing given the version being used.

danparizher commented:
Thanks for the detailed review!

I went ahead and version-gated magic globals by introducing `python_magic_globals` (with `__annotate__` behind `PY314_PLUS_MAGIC_GLOBALS`) and updated the linter to bind them per target version; I left `is_python_builtin` as-is since `__annotate__` isn’t a builtin.

I also reworked the pyflakes tests to explicitly check `__annotate__` on 3.13 vs 3.14, and added/adjusted ty mdtests to model `__warningregistry__` as possibly unbound and to validate the `__annotate__` gating. All tests pass locally.

ntBre reviewed (commented):
Thanks! This looks good to me on the Ruff side.
